File: Python/CManifold.py
import numpy as np
import time
from datetime import datetime

from output_log import output_txt_path


def _pyd_read(obj, attr):
    a = getattr(obj, attr, None)
    if a is None:
        return None
    return a() if callable(a) else a


class CManifold:

    def __init__(self, Lboard, add_syr):
        
        self.Lboard = Lboard
        self.add_syr = add_syr
        
        # General info
        self.device = []
        self.name = []
        self.address = []
        
        # Flags
        # self.FlagIsMoving = False
        self.FlagIsDone = True
        self.FlagIsOnline = False
        self.FlagIsMoving = False
        self.FlagIsStuck = False
        self.FlagReady = True
        self.V_missing = [False, False, False, False]
        self.V_status = [None, None, None, None]

        # Clocks
        self.ClockStartCmd = None
        self.ClockStopCmd = None

        ### Constructor
        self.device = self.Lboard.eib.New4VM(np.int8(self.add_syr))
        self.name= self.device.GetName()
        
        self.UpdateStatus()

        with open(output_txt_path(), "a") as OUTPUT:
            comment = f"Manifold {self.name} loaded."
            OUTPUT.write(comment + "\n")
            print(comment)

    ### UpdateStaus
    def UpdateStatus(self):
        self.device.CmdGetStatus()
        self.FlagIsDone = self.device.IsDone()
        self.FlagIsOnline = self.device.IsOnline()
        try:
            self.FlagIsMoving = self.device.IsMoving()
        except Exception:
            self.FlagIsMoving = False
        try:
            self.FlagIsStuck = self.device.IsStuck()
        except Exception:
            self.FlagIsStuck = False
        for i in range(1, 5):
            try:
                self.V_missing[i - 1] = bool(_pyd_read(self.device, f"V{i}Missing"))
            except Exception:
                self.V_missing[i - 1] = False
            try:
                self.V_status[i - 1] = _pyd_read(self.device, f"V{i}Status")
            except Exception:
                self.V_status[i - 1] = None

    def _switch_context(self, v1, v2, v3, v4):
        addr = getattr(self, "add_man", None)
        if addr is None:
            addr = getattr(self, "add_syr", None)
        if addr is None:
            addr = getattr(self, "address", "?")
        return (
            f"name={self.name}, addr={addr}, online={self.FlagIsOnline}, "
            f"ready={self.FlagReady}, v1={v1}, v2={v2}, v3={v3}, v4={v4}"
        )

    ### Switch Valves
    def SwitchValves(self, v1, v2, v3, v4):
        """CmdSetValves: 0=no change, 1=position A, 2=closed, 3=position B (uProcess).

        Raises RuntimeError if the manifold is offline / stuck at entry, or if
        the driver command returns False. Polls hardware until done with
        cooperative Stop-button checks (via parent board's cancel_requested /
        poll_hook) so the GUI doesn't freeze during valve motion.
        """
        try:
            self.UpdateStatus()
        except Exception as e:
            raise RuntimeError(
                f"Manifold {self.name}: UpdateStatus failed before SwitchValves: {e}"
            ) from e
        if not self.FlagIsOnline:
            raise RuntimeError(f"Manifold {self.name} is offline.")
        if getattr(self, "FlagIsStuck", False):
            raise RuntimeError(f"Manifold {self.name} is stuck.")

        rv = self.device.CmdSetValves(
            np.int8(v1), np.int8(v2), np.int8(v3), np.int8(v4)
        )
        if rv is False:
            raise RuntimeError(
                f"Manifold ({self._switch_context(v1, v2, v3, v4)}): CmdSetValves({v1},{v2},{v3},{v4}) returned False."
            )
        self.FlagReady = False
        self.displayswitch(v1, v2, v3, v4)

        start = time.monotonic()
        # Valve switches should complete in well under a second. 30s is a
        # generous ceiling that still prevents a dead manifold from hanging
        # the GUI thread forever.
        timeout_seconds = 30
        while self.FlagIsDone == False:
            if getattr(self.Lboard, "cancel_requested", False) or \
               getattr(self.Lboard, "Stop", False):
                break
            hook = getattr(self.Lboard, "poll_hook", None)
            if callable(hook):
                try:
                    hook()
                except Exception:
                    pass
            self.UpdateStatus()
            if getattr(self, "FlagIsStuck", False):
                raise RuntimeError(
                    f"Manifold {self.name} got stuck during switch."
                )
            if time.monotonic() - start > timeout_seconds:
                raise RuntimeError(
                    f"Manifold {self.name} switch timed out after "
                    f"{timeout_seconds}s."
                )
            time.sleep(0.01)
        if self.FlagIsDone == True:
            self.displayswitchstop()

    def SetValvesNative(self, v1, v2, v3, v4):
        """Set all four valves with native semantics (0–3); waits until done."""
        self.SwitchValves(v1, v2, v3, v4)

    def SetSingleValveMotion(self, valve_index_1_to_4: int, motion_code: int) -> bool:
        """CmdSetValveMotion: valve 1=V1…; motion uses same 0–3 scheme as CmdSetValves where applicable."""
        return bool(self.device.CmdSetValveMotion(int(valve_index_1_to_4), int(motion_code)))

    ### Display switch start
    def displayswitch(self,v1,v2,v3,v4):
        self.ClockStartCmd = datetime.now()
        self.UpdateStatus()
        with open(output_txt_path(), "a") as OUTPUT:
            comment = f"{self.ClockStartCmd.strftime('%X')} 4VM {self.name} is switching valves to {v1}, {v2}, {v3}, {v4}."
            OUTPUT.write(comment + "\n")
            print(comment)

    ### Display switch stop
    def displayswitchstop(self):
        self.ClockStopCmd = datetime.now()
        # self.UpdateStatus()
        with open(output_txt_path(), "a") as OUTPUT:
            comment = f"{self.ClockStopCmd.strftime('%X')} 4VM {self.name} is done."
            OUTPUT.write(comment + "\n")
            print(comment)
        self.FlagReady = True

File: Matlab/CManifold.m
classdef CManifold < handle    
    properties (GetAccess = 'public', SetAccess = 'public', SetObservable)
        // General info
        device = [];
        name = [];
        address = [];
        
        // Flags
//         FlagIsMoving = false;
        FlagIsDone = true;
        FlagIsOnline = false; 
        FlagReady = true;
        
        // Clocks
        ClockStartCmd
        ClockStopCmd
    end
    
     methods(Access = public)
         
        // Constructor
        function obj = CManifold(Lboard,add_syr)
            obj.device=Lboard.eib.New4VM(int8(add_syr));
            obj.name=char(obj.device.GetName);
            
            UpdateStatus(obj);
            
            diary on
            comment=['Manifold ' , obj.name, ' loaded.'];
            disp(comment);  
            diary off
        end
        
        // UpdateStaus
        function UpdateStatus(obj)
            obj.device.CmdGetStatus();
            obj.FlagIsDone=obj.device.IsDone();
            obj.FlagIsOnline=obj.device.IsOnline();                  
        end
        
                // Switch Valves
        function SwitchValves(obj,v1,v2,v3,v4)
            obj.device.CmdSetValves(int8(v1),int8(v2),int8(v3),int8(v4));
            obj.FlagReady = false;
            displayswitch(obj,v1,v2,v3,v4);
            while obj.FlagIsDone == false
                UpdateStatus(obj);
            end
            if obj.FlagIsDone == true
                displayswitchstop(obj)
            end
        end
        
        // Display switch start
        function displayswitch(obj,v1,v2,v3,v4)
            obj.ClockStartCmd=clock;
            UpdateStatus(obj);
            diary on
            comment=[num2str(obj.ClockStartCmd(4)) , ':' , num2str(obj.ClockStartCmd(5)) ,':' ,num2str(obj.ClockStartCmd(6)), ' 4VM ' , obj.name, ' is switching valves to ',num2str(v1) , ', ',num2str(v2) , ', ',num2str(v3) , ', ',num2str(v4) , '.'];  
            disp(comment);
            diary off
        end
        
        // Display switch stop
        function displayswitchstop(obj)
            obj.ClockStopCmd=clock;
/             UpdateStatus(obj);
            diary on
            comment=[num2str(obj.ClockStopCmd(4)) , ':' , num2str(obj.ClockStopCmd(5)) ,':' ,num2str(obj.ClockStopCmd(6)), ' 4VM ' , obj.name, ' is done.']; 
            disp(comment);
            diary off
            obj.FlagReady = true;
        end
     end
end